luigi/contrib/kubernetes.py

```diff
@@ -181,27 +211,90 @@ def signal_complete(self):
+                    self.__logger.info(l)
+
+    def __print_kubectl_hints(self):
+        self.__logger.info("To stream Pod logs, use:")
+        for pod in self.__get_pods():
+            self.__logger.info("`kubectl logs -f pod/%s`" % pod.name)
+
+    def __verify_job_has_started(self):
+        """Asserts that the job has successfully started"""
+        # Verify that the job started
+        self.__get_job()
+
+        # Verify that the pod started
+        pods = self.__get_pods()
+
+        assert len(pods) > 0, "No pod scheduled by " + self.uu_name
+        for pod in pods:
+            status = pod.obj['status']
+            for cont_stats in status.get('containerStatuses', []):
+                if 'terminated' in cont_stats['state']:
+                    t = cont_stats['state']['terminated']
+                    err_msg = "Pod %s %s (exit code %d). Logs: `kubectl logs pod/%s`" % (
+                        pod.name, t['reason'], t['exitCode'], pod.name)
+                    assert t['exitCode'] == 0, err_msg
+
+                if 'waiting' in cont_stats['state']:
+                    wr = cont_stats['state']['waiting']['reason']
+                    assert wr == 'ContainerCreating', "Pod %s %s. Logs: `kubectl logs pod/%s`" % (
+                        pod.name, wr, pod.name)
+
+            for cond in status['conditions']:
+                if 'message' in cond:
+                    if cond['reason'] == 'ContainersNotReady':
+                        return False
+                    assert cond['status'] != 'False', \
+                        "[ERROR] %s - %s" % (cond['reason'], cond['message'])
+        return True
+
     def __get_job_status(self):
         """Return the Kubernetes job status"""
-        # Look for the required job
-        jobs = Job.objects(self.__kube_api).filter(selector="luigi_task_id="
-                                                            + self.job_uuid)
-        # Raise an exception if no such job found
-        if len(jobs.response["items"]) == 0:
-            raise RuntimeError("Kubernetes job " + self.uu_name + " not found")
         # Figure out status and return it
-        job = Job(self.__kube_api, jobs.response["items"][0])
-        if ("succeeded" in job.obj["status"] and job.obj["status"]["succeeded"] > 0):
-            job.scale(replicas=0)  # Downscale the job, but keep it for logging
-            return "succeeded"
-        if ("failed" in job.obj["status"]):
+        job = self.__get_job()
+
+        if "succeeded" in job.obj["status"] and job.obj["status"]["succeeded"] > 0:
+            job.scale(replicas=0)
+            if self.print_pod_logs_on_exit:
+                self.__print_pod_logs()
+            if self.delete_on_success:
+                for pod in self.__get_pods():
+                    self.__logger.info("Deleting Pod " + pod.name)
+                    pod.delete()
+                self.__logger.info("Deleting Job " + job.name)
+                job.delete()
```

Comment:
This cleanup loop seems racy/unreliable in case of API errors. The apiserver can handle cleanup of the dependent objects by itself, if told to do so (documented at https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/). The way to do it here would be to pass `{"kind":"DeleteOptions","apiVersion":"v1","propagationPolicy":"Foreground"}` to the job deletion request, like this:
```py
delete_options_cascade = {"kind": "DeleteOptions", "apiVersion": "v1", "propagationPolicy": "Foreground"}
r = self.__kube_api.delete(json=delete_options_cascade, **job.api_kwargs())
if r.status_code != 404:
    self.__kube_api.raise_for_status(r)
```

Reply:
Fair point. Why did you pick `"propagationPolicy":"Foreground"`? Was thinking of using background cascading deletion. Will fix.